File: Python3/프로그래머스/2/155651. 호텔 대실/호텔 대실.py
import heapq
def time_to_min(time_str):
    hour, minute = map(int, time_str.split(":"))
    return hour*60+minute

def solution(book_time):
    answer = 0
    book_time.sort(key= lambda x:x[0])
    rooms = []
    
    for start_str, end_str in book_time:
        start = time_to_min(start_str)
        end = time_to_min(end_str) + 10  
        if rooms and rooms[0] <= start:
            heapq.heappop(rooms)
        else:
            answer+=1
            
        heapq.heappush(rooms, end)
            
        
    return answer

File: C++/프로그래머스/2/155651. 호텔 대실/호텔 대실.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

bool cmp(vector<string> a, vector<string> b){
    return a[0]<b[0];
}

int timeToMinute(string time) {
    int hour = stoi(time.substr(0, 2));
    int min = stoi(time.substr(3, 2));
    return hour * 60 + min;
}

int solution(vector<vector<string>> book_time) {
    int answer = 0;
    sort(book_time.begin(), book_time.end(), cmp);
    vector<bool>check(book_time[0].size(),false);
    priority_queue<int, vector<int>, greater<int>> pq;
    for(auto book: book_time){
        int start = timeToMinute(book[0]);
    int end = timeToMinute(book[1]) + 10;

    if (pq.empty() || pq.top() > start) {
        answer++; 
        pq.push(end);
    } else {
        pq.pop();
        pq.push(end);
    }
    }
    return answer;
}